x/ibc/02-client/client/rest/query.go

```diff
@@ -0,0 +1,185 @@
+package rest
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/gorilla/mux"
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmtypes "github.com/tendermint/tendermint/types"
+
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/types/rest"
+	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/client/utils"
+	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/types/tendermint"
+	commitment "github.com/cosmos/cosmos-sdk/x/ibc/23-commitment"
+)
+
+func registerQueryRoutes(cliCtx context.CLIContext, r *mux.Router, queryRoute string) {
+	r.HandleFunc(fmt.Sprintf("/ibc/clients/{%s}/consensus-state", RestClientID), queryConsensusStateHandlerFn(cliCtx, queryRoute)).Methods("GET")
+	r.HandleFunc(fmt.Sprintf("/ibc/clients/{%s}/client-state", RestClientID), queryClientStateHandlerFn(cliCtx, queryRoute)).Methods("GET")
+	r.HandleFunc(fmt.Sprintf("/ibc/clients/{%s}/roots/{%s}", RestClientID, RestRootHeight), queryRootHandlerFn(cliCtx, queryRoute)).Methods("GET")
+	r.HandleFunc("/ibc/header", queryHeaderHandlerFn(cliCtx)).Methods("GET")
+	r.HandleFunc("/ibc/node-state", queryNodeConsensusStateHandlerFn(cliCtx)).Methods("GET")
+	r.HandleFunc("/ibc/path", queryPathHandlerFn(cliCtx)).Methods("GET")
+}
+
+func queryConsensusStateHandlerFn(cliCtx context.CLIContext, queryRoute string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		clientID := vars[RestClientID]
+
+		cliCtx, ok := rest.ParseQueryHeightOrReturnBadRequest(w, cliCtx, r)
+		if !ok {
+			return
+		}
+
+		req := abci.RequestQuery{
+			Path:  "store/ibc/key",
+			Data:  types.KeyConsensusState(clientID),
+			Prove: rest.ParseQueryProve(r),
+		}
+
+		res, err := cliCtx.QueryABCI(req)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+
+		var cs tendermint.ConsensusState
+		if err := cliCtx.Codec.UnmarshalBinaryLengthPrefixed(res.Value, &cs); err != nil {
+			rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+		}
+
+		cliCtx = cliCtx.WithHeight(res.Height)
+		rest.PostProcessResponse(w, cliCtx, types.NewConsensusStateResponse(clientID, cs, res.Proof, res.Height))
+	}
+}
+
+func queryHeaderHandlerFn(cliCtx context.CLIContext) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		header, err := utils.GetTendermintHeader(cliCtx)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+
+		rest.PostProcessResponse(w, cliCtx, header)
+	}
+}
+
+func queryClientStateHandlerFn(cliCtx context.CLIContext, queryRoute string) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		vars := mux.Vars(r)
+		clientID := vars[RestClientID]
+
+		cliCtx, ok := rest.ParseQueryHeightOrReturnBadRequest(w, cliCtx, r)
+		if !ok {
+			return
+		}
+
+		req := abci.RequestQuery{
+			Path:  "store/ibc/key",
```

Comment:
why are we querying the store directly instead of using the Querier? the latter provides better error messages and uses the standards of the SDK for creating REST endpoints.

Reply:
As I know, we can't get the Merkel proof using the Querier, so we query the store directly.

Reply:
you can retrieve the proof, you can use `res.Proof`. Check the `client/cli/query.go` references on the ibc module

Reply:
I believe we can't get the proof unless we modify the Querier function. Cli query cannot return proof if using Querier, as described by this issue: #5317.

Reply:
ok, thanks. Let's leave it as is here and we can fix the other issue next